File: src/main/java/DataStructureAlgo/Java/LeetCode2025/ProblemSet/Trees/pathSum/_666/PthSumIV_666.java
package DataStructureAlgo.Java.LeetCode2025.ProblemSet.Trees.pathSum._666;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

/**
 * Author: Nitin Gupta
 * Date: 8/24/2024
 * Question Category: 666 - Path Sum IV
 * Description: https://leetcode.com/problems/path-sum-iv/ && https://leetcode.ca/2017-09-26-666-Path-Sum-IV/
 * <p>
 * If the depth of a tree is smaller than 5, then this tree can be represented by an array of three-digit integers. For each integer in this array:
 * <p>
 * The hundreds digit represents the depth d of this node where 1 <= d <= 4.
 * The tens digit represents the position p of this node in the level it belongs to where 1 <= p <= 8. The position is the same as that in a full binary tree.
 * The units digit represents the value v of this node where 0 <= v <= 9.
 * Given an array of ascending three-digit integers nums representing a binary tree with a depth smaller than 5, return the sum of all paths from the root towards the leaves.
 * <p>
 * It is guaranteed that the given array represents a valid connected binary tree.
 * <p>
 * <p>
 * <p>
 * Example 1:
 * <p>
 * <p>
 * <p>
 * Input: nums = [113,215,221]
 * Output: 12
 * Explanation: The tree that the list represents is shown.
 * The path sum is (3 + 5) + (3 + 1) = 12.
 * Example 2:
 * <p>
 * <p>
 * <p>
 * Input: nums = [113,221]
 * Output: 4
 * Explanation: The tree that the list represents is shown.
 * The path sum is (3 + 1) = 4.
 * <p>
 * <p>
 * Constraints:
 * <p>
 * 1 <= nums.length <= 15
 * 110 <= nums[i] <= 489
 * nums represents a valid binary tree with depth less than 5.
 * <p>
 * <p>
 * File reference
 * -----------
 * Duplicate {@link}
 * Similar {@link}
 * extension {@link }
 * <p>
 * Tags
 * -----
 *
 * @medium
 * @BinaryTree
 * @Tree
 * @DepthFirstSearch
 * @LeetCodeLockedProblem
 * @PremimumQuestion <p>
 * <p>
 * Company Tags
 * -----
 * @Alibaba
 * @Editorial
 */
public class PthSumIV_666 {


    public static void main(String[] args) {

        boolean test = true;
        test &= test(new int[]{113, 215, 221}, 12);
        test &= test(new int[]{113, 221}, 4);
        System.out.println((test ? "All passed" : "Something Failed"));
    }

    private static boolean test(int[] nums, int expected) {
        System.out.println("-----------------");
        System.out.println(" Input : " + Arrays.toString(nums) + " expected : " + expected);

        Solution obj = new Solution();
        int actual = obj.pathSum(nums);
        System.out.println(" Output : " + actual + (actual == expected ? " Passed " : " Failed "));
        return actual == expected;
    }

    static class Solution {
        class Pair {
            int depth;
            int position;

            Pair(int depth, int position) {
                this.depth = depth;
                this.position = position;
            }

            @Override
            public boolean equals(Object o) {
                if (this == o) return true;
                if (o == null || getClass() != o.getClass()) return false;
                Pair pair = (Pair) o;
                return depth == pair.depth &&
                        position == pair.position;

            }

            @Override
            public int hashCode() {
                return Objects.hash(depth, position);
            }
        }

        /**
         * https://medium.com/shuwens-leetcode-solution/leetcode-666-path-sum-iv-2fecddf78253
         *
         * @param nums
         * @return
         */

        public int pathSum(int[] nums) {

            if (nums == null)
                return 0;

            final Map<Pair, Integer> depthPositionVsValue = new HashMap<>();
            for (int num : nums) {
                int depth = num / 100;
                int position = (num - depth * 100) / 10;
                int value = num % 10;

                Pair pair = new Pair(depth - 1, position - 1); // to make it 0 based indexed
                depthPositionVsValue.put(pair, value);
            }
            int[] total = {0};
            dfs(depthPositionVsValue, 0, 0, 0, total);
            return total[0];

        }

        private void dfs(Map<Pair, Integer> map, int depth, int pos, int currentSum, int[] total) {

            Integer root = map.get(new Pair(depth, pos));

            if (root == null)
                return;

            Integer leftChild = map.get(new Pair(depth + 1, pos * 2));
            Integer rightChild = map.get(new Pair(depth + 1, pos * 2 + 1));

            //left and right exist, go deep
            if (leftChild != null && rightChild != null) {

                dfs(map, depth + 1, pos * 2, currentSum + root, total);
                dfs(map, depth + 1, pos * 2 + 1, currentSum + root, total);

            } else if (leftChild != null) {

                dfs(map, depth + 1, pos * 2, currentSum + root, total);

            } else if (rightChild != null) {

                dfs(map, depth + 1, pos * 2 + 1, currentSum + root, total);

            } else {
                // leaf node
                currentSum += root;
                total[0] += currentSum;
                return;

            }
        }
    }

}

File: src/main/java/DataStructureAlgo/Java/LeetCode2025/ProblemSet/Trees/pathSum/_666/PthSumIV_666.py
"""
Author: Nitin Gupta
Date: 6/20/2025
Question Category: 666 - Path Sum IV
Description: https://leetcode.com/problems/path-sum-iv/ && https://leetcode.ca/2017-09-26-666-Path-Sum-IV/
Description:
File reference
-----------
Duplicate {@link}
Similar {@link}
extension {@link }
DP-BaseProblem {@link }
<p><p>
Tags
-----

<p><p>
Company Tags
-----
<p>
-----

@Editorial <p><p>
-----
@OptimalSolution {@link }
"""




class Solution:
    def __init__(self):
        self.all_path_sum = 0
        self.cache : Dict[Tuple[int,int], int] = {}

    def pathSum(self, nums: List[int]) -> int:
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return 0
        
        #build cache for traversing the tree
        #for each node in a binary tree; root is at position p, left child is at position 2*p, right child is at position 2*p+1
        # cache will store the [depth,post] -> value
        self.all_path_sum: int = 0
        
        for num in nums:
            #extract the depth of the node, the depth is the first digit of the number
            depth = num // 100
            
            #extract the position of the node, the position is the second and third digits of the number
            position = (num - depth * 100) // 10
            
            #extrac the value of the node, the value is the last digit of the number
            value = num % 10
            
            self.cache[(depth-1,position-1)] = value #depth and pos would be 0 indexed
            
        
        def dfs(depth: int, position: int, current_sum: int) -> None:
            #extract the value of the node at [depth,position]
            root = self.cache.get((depth,position))
            
            #if a tree is complete, return
            if not root:
                return  
            
            #run its left and right child
            left_child = self.cache.get((depth+1, 2*position))
            right_child = self.cache.get((depth+1, 2*position+1))

            if left_child:
                dfs(depth+1, 2*position, current_sum+root)
            
            if right_child:
                dfs(depth+1, 2*position+1, current_sum+root)

            #this is the leaf node, get the path sum
            if not left_child and not right_child:
                current_sum+= root
                self.all_path_sum += current_sum  
        
        dfs(depth=0, position=0, current_sum=0)
        return self.all_path_sum
            
            
        

def test(input_data, expected):
    """
    Test function to verify the solution
    """
    CommonMethods.print_test(["Input", "Expected"], True, input_data, expected)
    pass_test, final_pass = True, True

    output = Solution().pathSum(input_data)

    pass_test = CommonMethods.compare_result(output, expected, True)
    CommonMethods.print_test(["Output", "Pass"], False, output, "PASS" if pass_test else "FAIL")
    final_pass &= pass_test

    return final_pass


if __name__ == "__main__":
    tests: List = [test([113,215,221], 12),
                   test([113,221], 4)
                   ]

    CommonMethods.print_all_test_results(tests)
